gameengine.js

```diff
@@ -114,9 +111,41 @@ class GameEngine {
                     that.A = false;
                     break;
             }
-        }, false);
+        }
+
+        that.mousemove = mouseListener;
+        that.leftclick = mouseClickListener;
+        that.wheelscroll = wheelListener;
+        that.keydown = keydownListener;
+        that.keyup = keyUpListener;
+
+        this.ctx.canvas.addEventListener("mousemove", that.mousemove, false);
+
+        this.ctx.canvas.addEventListener("click", that.leftclick, false);
+
+        this.ctx.canvas.addEventListener("wheel", that.wheelscroll, false);
+
+        this.ctx.canvas.addEventListener("keydown", that.keydown, false);
+
+        this.ctx.canvas.addEventListener("keyup", that.keyup, false);
     };
 
+    disableInput() {
```

Comment:
Maybe move to a flag based system rather than a function manipulating listeners? Like have a `isInputEnabled` property on the game engine and check that property in the input functions above (starting line 46). This way you never have to worry about duplicate listeners or making sure you called startInput again. It reduces the amount of states you have to think about. Optional recommendation.